docs/usage/data-modeling.md

```diff
@@ -0,0 +1,150 @@
+`user_data`'s contents are arbitrary, interpreted only by the application.
+
+`user_data` is indexed for efficient point and range queries.
+
+Example uses:
+
+- Set `user_data` to a "foreign key" — that is, an identifier of a corresponding object within
+  another database.
+- Set `user_data` to a group identifier for objects that will be queried together.
+- Set `user_data` to a transfer or account `id`.
+  (TODO: Can we use this for join queries via the query API, or must the application implement them?)
+
+## `id`
+
+The primary purpose of an `id` (for both [accounts](../reference/accounts.md#id) and
+[transfers](../reference/transfers.md#id)) is to serve as an "idempotency key".
+The database uses ids to avoid repeating duplicate events.
+
+[Randomly-generated identifiers](#random-identifier) are recommended for most applications.
+
+When selecting an `id` scheme:
+
+- Idempotency is particularly important (and difficult) in the context of
+  [application crash recovery](./integration.md#consistency-with-foreign-databases).
+- Be careful to [avoid `id` collisions](https://en.wikipedia.org/wiki/Birthday_problem).
+- An account and a transfer may share the same `id` — they belong to different "namespaces".
+- Avoid requiring a central oracle to generate each unique `id` (e.g. an auto-increment field in SQL).
+  A central oracle may become a performance bottleneck when creating accounts/transfers.
+- `id` is mostly accessed by point queries, but it is indexed for efficient range queries as well.
+
+### Examples
+#### Random Identifier
+
+Randomly-generated identifiers are recommended for most applications.
+
+- Random identifiers require coordination with a secondary database to implement idempotent
+  [application crash recovery](./integration.md#consistency-with-foreign-databases).
+- Random identifiers have an insignificant risk of collisions.
+- Random identifiers do not require a central oracle.
+- Only point queries are useful for fetching randomly-generated identifiers.
+
+To maximize id entropy, prefer a cryptographically-secure PRNG.
+
+#### Reuse Foreign Identifier
+
+This technique is most appropriate when integrating TigerBeetle with an existing application
+where TigerBeetle accounts or transfers map one-to-one with an entity in the foreign database.
+
+Set `id` to a "foreign key" — that is, reuse an identifier of a corresponding object from another
+database. For example, if every user (within the application's database) has a single account, then
+the identifier within the foreign database can be used as the `Account.id` within TigerBeetle.
+
+To reuse the foreign identifier, it must conform to TigerBeetle's `id`
+[constraints](../reference/accounts.md#id).
+
+Like [randomly-generated identifiers](#random-identifier), this technique requires careful
+coordination with the foreign database for idempotent
+[application crash recovery](./integration.md#consistency-with-foreign-databases).
+
+#### Logically Grouped Objects
```

Comment:
Maybe we should add `(Advanced)` to the headers of these that are not random ids? Just so that we all the more encourage people to use random ids

Reply:
:thinking: I'd rather not put it in the header because it pollutes the URL fragment. I think it is ok; there are 2 different places that say "Randomly-generated identifiers are recommended for most applications.".

Reply:
It's just that by being equal weight headers they otherwise look like equally suggested

Reply:
How about two different "Examples" sections; "Examples" and "Examples (Advanced)"?

Reply:
Sure :+1: